### PR TITLE
[FIX] website_sale: fix product image reordering issues

### PR DESCRIPTION
**Pull request by ltinel**
This change fixes the unexpected behavior of product image reordering:

1. Previously, when moving an image to the first or last postion, it was swapped with the first or last image. Now, it is inserted in the first or last postion, while keeping the relative ordering of the other images unchanged.

2. Previously, the main image could be in any position, but as soon as it was reordered, it would jump to the first position. Now, the main image is always in first position.

task-3581895

### CODE REVIEW

File: addons/website_sale/controllers/main.py [hunk 1]
```diff
@@ -577,28 +575,38 @@ def resequence_product_image(self, image_res_model, image_res_id, move):
...
+            if additional_image.video_url:
+                raise ValidationError(_("You can't use a video as the product's main image."))
+            # Swap image data.
+            main_image.image_1920, additional_image.image_1920 = (
+                additional_image.image_1920, main_image.image_1920
+            )
+            additional_image.name = main_image.name  # Update image name but not product name.
+            # Swap records.
+            product_images[main_image_idx], product_images[0] = (
+                product_images[0], product_images[main_image_idx]
```
Comment (vchu-odoo) [inline addons/website_sale/controllers/main.py:603]:
```suggestion
                additional_image, main_image
```

File: addons/website_sale/tests/test_website_editor.py [hunk 2]
```diff
@@ -116,79 +122,121 @@ def test_extra_images_with_new_variant(self):
             )
         self.assertEqual(1, len(product_template.product_variant_ids))
 
-    def test_resequence_images(self):
+    def test_resequence_image_first(self):
```
Comment (vchu-odoo) [inline addons/website_sale/tests/test_website_editor.py:125]: maybe you can add an assert to check if the `i3` became the main image of the product.

File: addons/website_sale/tests/test_website_editor.py [hunk 3]
```diff
@@ -116,79 +122,121 @@ def test_extra_images_with_new_variant(self):
...
+            i1, i2, i3, i4, i5, i6 = self._get_product_image_data()
             self.WebsiteSaleController.resequence_product_image(
                 images[2]._name,
                 images[2].id,
                 'last',
             )
+            self.env['product.image'].invalidate_model()
+            self.assertListEqual(self._get_product_image_data(), [i1, i2, i4, i5, i6, i3])
+
+    def test_resequence_image_first_to_last(self):
```
Comment (vchu-odoo) [inline addons/website_sale/tests/test_website_editor.py:178]: same check can be added here.

**Review by vchu-odoo — APPROVED** [review]
Hello,
Thank you for such a great work on this task!

I approve the PR since there is no blocking issues just small suggestions.

File: addons/website_sale/tests/test_website_editor.py (hunk 2 above)
Comment (ltinel) [reply addons/website_sale/tests/test_website_editor.py:125]: Good idea, done :)

File: addons/website_sale/tests/test_website_editor.py (hunk 3 above)
Comment (ltinel) [reply addons/website_sale/tests/test_website_editor.py:178]: Done

File: addons/website_sale/controllers/main.py (hunk 1 above)
Comment (ltinel) [reply addons/website_sale/controllers/main.py:603]: Done.

File: addons/website_sale/controllers/main.py [hunk 4]
```diff
@@ -577,28 +575,38 @@ def resequence_product_image(self, image_res_model, image_res_id, move):
...
+
+        # If the main image has been reordered (i.e. it's no longer in first position), use the
+        # image that's now in first position as main image instead.
+        main_image_idx = 0
+        for idx, product_image in enumerate(product_images):
+            # Additional images are product.image records. The main image is a product.product or
+            # product.template record.
+            if product_image._name != 'product.image':
+                main_image_idx = idx
+                break
```
[the remark this answers is not in the document]
Comment (ltinel) [reply addons/website_sale/controllers/main.py:590]: I don't mind doing this, but this will systematically look at all images, which can be a bit slower. Also, at this point in the code, there should always be a main image, so we don't need the default, and we could simply write `[idx for idx, image in enumerate(product_images) if image._name != 'product.image'][0]`. WDYT?

[the remark this answers is not in the document]
Comment (vchu-odoo) [reply addons/website_sale/controllers/main.py:590]: As far as I know `next` returns the first element that matches the condition, so it is equivalent of `for` loop with `break`. In your suggestion it will always go through the whole list.
The default can be removed if there is always a main image ofc.

[the remark this answers is not in the document]
Comment (ltinel) [reply addons/website_sale/controllers/main.py:590]: Oh, so the first argument in the next function is an iterator? I thought you were passing a list (evaluated before applying next). Either way, it doesn't change the complexity so I'll update it.

File: addons/website_sale/tests/test_website_editor.py [hunk 5]
```diff
@@ -116,79 +122,123 @@ def test_extra_images_with_new_variant(self):
...
+            i1, i2, i3, i4, i5, i6 = self._get_product_image_data()
             self.WebsiteSaleController.resequence_product_image(
                 images[2]._name,
                 images[2].id,
                 'first',
             )
+            # Trigger the reordering of product.image records based on their sequence.
+            self.env['product.image'].invalidate_model()
+            self.assertListEqual(self._get_product_image_data(), [i3, i1, i2, i4, i5, i6])
+            self.assertEqual(self.product.image_1920, i3)
```
[the remark this answers is not in the document]
Comment (ltinel) [reply addons/website_sale/tests/test_website_editor.py:138]: I must admit I'm not a big fan of adding custom messages to test assertions, so I'd rather keep it as is if that's ok with you?

[the remark this answers is not in the document]
Comment (vchu-odoo) [reply addons/website_sale/tests/test_website_editor.py:138]: Usually it is useful to understand what is being tested but if you think that it is self-explanatory and not needed then fine by me.

File: addons/website_sale/controllers/main.py [hunk 6]
```diff
@@ -546,8 +546,6 @@ def resequence_product_image(self, image_res_model, image_res_id, move):
```
Comment (AntoineVDV) [inline addons/website_sale/controllers/main.py:546]: It could be worth to add a docstring to the method:
```py
def resequence_product_image(self, image_res_model, image_res_id, move):
    """ Move the product image in the given direction and update all images' sequence.

    :param str image_res_model: The model of the image. It can be 'product.template',
                                'product.product', or 'product.image'.
    :param str image_res_id: The record ID of the image to move.
    :param str move: The direction of the move. It can be 'first', 'left', 'right', or 'last'.
    :raises NotFound: If the user does not have the required permissions, if the model of the
                      image is not allowed, or if the move direction is not allowed.
    :raise ValidationError: If the product is not found.
    :raise ValidationError: If the image to move is not found in the product images.
    :return: None
    """
```

File: addons/website_sale/tests/test_website_editor.py [hunk 7]
```diff
@@ -116,79 +122,123 @@ def test_extra_images_with_new_variant(self):
...
         with MockRequest(self.product.env, website=self.website):
-            # Test moving to first position
             images = self.product._get_images()
-            data_source = images[2].image_1920
-            data_target = images[0].image_1920
+            i1, i2, i3, i4, i5, i6 = self._get_product_image_data()
             self.WebsiteSaleController.resequence_product_image(
                 images[2]._name,
                 images[2].id,
                 'first',
```
Comment (AntoineVDV) [inline addons/website_sale/tests/test_website_editor.py:133]:
```suggestion
                images[2]._name, images[2].id, 'first'
```

File: addons/website_sale/tests/test_website_editor.py [hunk 8]
```diff
@@ -116,79 +122,123 @@ def test_extra_images_with_new_variant(self):
...
-            self.assertEqual(images[2].image_1920, data_target)
-            self.assertEqual(images[0].image_1920, data_source)
-            # Test moving one to the left
-            data_source = images[2].image_1920
-            data_target = images[1].image_1920
+            i1, i2, i3, i4, i5, i6 = self._get_product_image_data()
             self.WebsiteSaleController.resequence_product_image(
                 images[2]._name,
                 images[2].id,
                 'left',
```
Comment (AntoineVDV) [inline addons/website_sale/tests/test_website_editor.py:148]:
```suggestion
                images[2]._name, images[2].id, 'left'
```

File: addons/website_sale/tests/test_website_editor.py [hunk 9]
```diff
@@ -116,79 +122,123 @@ def test_extra_images_with_new_variant(self):
...
-            self.assertEqual(images[2].image_1920, data_target)
-            self.assertEqual(images[1].image_1920, data_source)
-            # Test moving one to the right
-            data_source = images[2].image_1920
-            data_target = images[3].image_1920
+            i1, i2, i3, i4, i5, i6 = self._get_product_image_data()
             self.WebsiteSaleController.resequence_product_image(
                 images[2]._name,
                 images[2].id,
                 'right',
```
Comment (AntoineVDV) [inline addons/website_sale/tests/test_website_editor.py:161]:
```suggestion
                images[2]._name, images[2].id, 'right'
```

File: addons/website_sale/tests/test_website_editor.py [hunk 10]
```diff
@@ -116,79 +122,123 @@ def test_extra_images_with_new_variant(self):
...
-            self.assertEqual(images[2].image_1920, data_target)
-            self.assertEqual(images[3].image_1920, data_source)
-            # Test moving one to the last position
-            data_source = images[2].image_1920
-            data_target = images[-1].image_1920
+            i1, i2, i3, i4, i5, i6 = self._get_product_image_data()
             self.WebsiteSaleController.resequence_product_image(
                 images[2]._name,
                 images[2].id,
                 'last',
```
Comment (AntoineVDV) [inline addons/website_sale/tests/test_website_editor.py:174]:
```suggestion
                images[2]._name, images[2].id, 'last'
```

File: addons/website_sale/tests/test_website_editor.py [hunk 11]
```diff
@@ -116,79 +122,123 @@ def test_extra_images_with_new_variant(self):
...
             )
+            self.env['product.image'].invalidate_model()
+            self.assertListEqual(self._get_product_image_data(), [i1, i2, i4, i5, i6, i3])
+
+    def test_resequence_image_first_to_last(self):
+        """
+        Moving an image from first to last position is an edge case.
+        This test ensures it won't be accidentally broken.
```
Comment (AntoineVDV) [inline addons/website_sale/tests/test_website_editor.py:182]: Why? It's a valid flow and the button is small enough that you can simply misclick it :sweat_smile: 

For once I find this comment useless, especially the second sentence.

File: addons/website_sale/tests/test_website_editor.py [hunk 12]
```diff
@@ -116,79 +122,123 @@ def test_extra_images_with_new_variant(self):
...
+        """
+        self._create_product_images()
+        with MockRequest(self.product.env, website=self.website):
+            images = self.product._get_images()
+            i1, i2, i3, i4, i5, i6 = self._get_product_image_data()
+            self.WebsiteSaleController.resequence_product_image(
+                images[0]._name,
+                images[0].id,
+                'last',
```
Comment (AntoineVDV) [inline addons/website_sale/tests/test_website_editor.py:191]:
```suggestion
                images[0]._name, images[0].id, 'last'
```

File: addons/website_sale/tests/test_website_editor.py [hunk 13]
```diff
@@ -116,79 +122,123 @@ def test_extra_images_with_new_variant(self):
...
-            self.assertEqual(images[2].image_1920, data_target)
-            self.assertEqual(images[-1].image_1920, data_source)
-            # Test moving an image with a video_url instead of image_1920
-            data_target = images[1].image_1920
             images[2].video_url = "https://www.youtube.com/watch?v=dQw4w9WgXcQ"
+            i1, i2, i3, i4, i5, i6 = self._get_product_image_data()
             self.WebsiteSaleController.resequence_product_image(
                 images[2]._name,
                 images[2].id,
                 'left',
```
Comment (AntoineVDV) [inline addons/website_sale/tests/test_website_editor.py:206]:
```suggestion
                images[2]._name, images[2].id, 'left'
```

File: addons/website_sale/tests/test_website_editor.py [hunk 14]
```diff
@@ -116,79 +122,123 @@ def test_extra_images_with_new_variant(self):
...
+            self.WebsiteSaleController.resequence_product_image(
+                images[0]._name,
+                images[0].id,
+                'last',
+            )
+            self.env['product.image'].invalidate_model()
+            self.assertListEqual(self._get_product_image_data(), [i2, i3, i4, i5, i6, i1])
+            self.assertEqual(self.product.image_1920, i2)
+
+    def test_resequence_video(self):
```
Comment (AntoineVDV) [inline addons/website_sale/tests/test_website_editor.py:197]:
```suggestion
    def test_resequence_video_left(self):
```

File: addons/website_sale/tests/test_website_editor.py [hunk 15]
```diff
@@ -116,79 +122,123 @@ def test_extra_images_with_new_variant(self):
...
+            images[2].video_url = "https://www.youtube.com/watch?v=dQw4w9WgXcQ"
+            i1, i2, i3, i4, i5, i6 = self._get_product_image_data()
             with self.assertRaises(ValidationError):
                 self.WebsiteSaleController.resequence_product_image(
-                    images[1]._name,
-                    images[1].id,
-                    'left',
+                    images[2]._name,
+                    images[2].id,
+                    'first',
```
Comment (AntoineVDV) [inline addons/website_sale/tests/test_website_editor.py:222]:
```suggestion
                    images[2]._name, images[2].id, 'first'
```

File: addons/website_sale/tests/test_website_editor.py [hunk 16]
```diff
@@ -116,79 +122,123 @@ def test_extra_images_with_new_variant(self):
...
+            images[1].video_url = "https://www.youtube.com/watch?v=dQw4w9WgXcQ"
+            i1, i2, i3, i4, i5, i6 = self._get_product_image_data()
             with self.assertRaises(ValidationError):
                 self.WebsiteSaleController.resequence_product_image(
-                    images[1]._name,
-                    images[1].id,
-                    'first',
+                    images[0]._name,
+                    images[0].id,
+                    'right',
```
Comment (AntoineVDV) [inline addons/website_sale/tests/test_website_editor.py:238]:
```suggestion
                    images[0]._name, images[0].id, 'right'
```

**Review by AntoineVDV — APPROVED** [review]
Good job! This code now seems reliable while it was dubious at best before :slightly_smiling_face: 

@robodoo delegate+ override=ci/security
Reason: `.open` - no change; the variable is simply renamed.
Link to ci/security build (16.0): https://runbot.odoo.com/runbot/build/59699249

File: addons/website_sale/tests/test_website_editor.py (hunk 14 above)
Comment (ltinel) [reply addons/website_sale/tests/test_website_editor.py]: Done (FYI the idea here was to have a test case for moving a video, not for a specific move, which is why I omitted the move type).

File: addons/website_sale/controllers/main.py (hunk 6 above)
Comment (ltinel) [reply addons/website_sale/controllers/main.py:560]: Done. I also documented the validation error which is throw when a video is moved to the first position.

**Review by ltinel — COMMENTED** [review]
Thanks for the review @AntoineVDV!

File: addons/website_sale/tests/test_website_editor.py (hunk 11 above)
Comment (ltinel) [reply addons/website_sale/tests/test_website_editor.py]: (FYI, I slightly adjusted the comment)

**Comment by ltinel** [comment]
robodoo r+
